presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -148,27 +216,51 @@ protected Page computeNext()
                         BlockBuilder output = pageBuilder.getBlockBuilder(types.size() + i);
                         aggregator.evaluate(groupId, output);
                     }
-
-                    groupId++;
                 }
 
                 return pageBuilder.build();
             }
         };
     }
 
+    public List<Type> buildTypes()
+    {
+        ArrayList<Type> types = new ArrayList<>(groupByHash.getTypes());
+        for (Aggregator aggregator : aggregators) {
+            types.add(aggregator.getType());
+        }
+        return types;
+    }
+
+    private Iterator<Integer> consecutiveGroupIds()
+    {
+        return IntStream.range(0, groupByHash.getGroupCount()).iterator();
+    }
+
+    private Iterator<Integer> hashSortedGroupIds()
+    {
+        List<Integer> groupIds = Lists.newArrayList(consecutiveGroupIds());
```

Comment:
The amount of system memory is configurable, so the user could easily set it to GBs.

I would fix this in one of two ways:
1) limit the number of groups spilled to a small constant
2) account for this array when accepting input, so that there is spare memory left to make this allocation.

Trying to drop memory from GroupByHash seems too complicated